Realm/ObjectStore/impl/background_collection.cpp

```diff
@@ -0,0 +1,832 @@
+        changeset.moves.push_back({row.prev_tv_index, row.tv_index});
+        changeset.insertions.add(row.tv_index);
+        removed.add(row.prev_tv_index);
+    }
+}
+
+class SortedMoveCalculator {
+public:
+    SortedMoveCalculator(std::vector<RowInfo>& new_rows, CollectionChangeIndices& changeset)
+    : m_modified(changeset.modifications)
+    {
+        std::vector<Row> old_candidates;
+        old_candidates.reserve(new_rows.size());
+        for (auto& row : new_rows) {
+            old_candidates.push_back({row.row_index, row.prev_tv_index});
+        }
+        std::sort(begin(old_candidates), end(old_candidates), [](auto a, auto b) {
+            return std::tie(a.tv_index, a.row_index) < std::tie(b.tv_index, b.row_index);
+        });
+
+        // First check if the order of any of the rows actually changed
+        size_t first_difference = npos;
+        for (size_t i = 0; i < old_candidates.size(); ++i) {
+            if (old_candidates[i].row_index != new_rows[i].row_index) {
+                first_difference = i;
+                break;
+            }
+        }
+        if (first_difference == npos)
+            return;
+
+        // A map from row index -> tv index in new results
+        b.reserve(new_rows.size());
+        for (size_t i = 0; i < new_rows.size(); ++i)
+            b.push_back({new_rows[i].row_index, i});
+        std::sort(begin(b), end(b), [](auto a, auto b) {
+            return std::tie(a.row_index, a.tv_index) < std::tie(b.row_index, b.tv_index);
+        });
+
+        a = std::move(old_candidates);
+
+        find_longest_matches(first_difference, a.size(),
+                             first_difference, new_rows.size());
+        m_longest_matches.push_back({a.size(), new_rows.size(), 0});
+
+        size_t i = first_difference, j = first_difference;
+        for (auto match : m_longest_matches) {
+            for (; i < match.i; ++i)
+                changeset.deletions.add(a[i].tv_index);
+            for (; j < match.j; ++j)
+                changeset.insertions.add(new_rows[j].tv_index);
+            i += match.size;
+            j += match.size;
+        }
+    }
+
+private:
+    struct Match {
+        size_t i, j, size, modified;
+    };
+    struct Row {
+        size_t row_index;
+        size_t tv_index;
+    };
+
+    IndexSet const& m_modified;
+    std::vector<Match> m_longest_matches;
+
+    std::vector<Row> a, b;
+
+    Match find_longest_match(size_t begin1, size_t end1, size_t begin2, size_t end2)
+    {
+        struct Length {
+            size_t j, len;
+        };
+        std::vector<Length> cur;
+        std::vector<Length> prev;
+
+        auto length = [&](size_t j) -> size_t {
+            for (auto const& pair : prev) {
+                if (pair.j + 1 == j)
+                    return pair.len + 1;
+            }
+            return 1;
+        };
+
+        Match best = {begin1, begin2, 0, 0};
+
+        for (size_t i = begin1; i < end1; ++i) {
```

Comment:
It's hard to see at a high level what this loop is doing. The generic variable names don't help. Can you add a comment explaining briefly what this loop is doing?